### PR TITLE
Addition of Okta and a generic OIDC client for authentication options

### PR DESCRIPTION
**Pull request by cginmn**
Using the Microsoft configurations and code, made an Okta and Generic OpenID Connect client.

Description about what this pull request does.

Using the Microsoft code and configuration, made an Okta and a generic OpenID Connect client. 
Create Okta and generic OIDC client endpoint config
fence/resources/openid/okta_oauth2.py
fence/resources/openid/generic_oauth2.py

Create Okta and generic methods
fence/blueprints/login/okta.py
fence/blueprints/login/generic.py

Update Okta and generic OIDC entry
fence/config-default.yaml

Update Okta and generic login route
fence/blueprints/login/__init__.py

Update Okta and generic OIDC client
fence/__init__.py

Please make sure to follow the [DEV guidelines](https://gen3.org/resources/developer/dev-introduction/) before asking for review.

### New Features
- Implemented new Okta and generic OIDC clients

### CODE REVIEW

**Comment by ohofmann** [comment]
A +1 here - we'd be interested in this.

File: fence/resources/openid/generic_oauth2.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+import json
```
Comment (williamhaley) [inline fence/resources/openid/generic_oauth2.py:1]: Unused import.

```suggestion
```

File: fence/blueprints/login/okta.py [hunk 2]
```diff
@@ -0,0 +1,21 @@
+import flask
+
+from fence.models import IdentityProvider
```
Comment (williamhaley) [inline fence/blueprints/login/okta.py:3]: Unused import.

```suggestion
```

File: fence/blueprints/login/generic.py [hunk 3]
```diff
@@ -0,0 +1,21 @@
+import flask
+
+from fence.models import IdentityProvider
```
Comment (williamhaley) [inline fence/blueprints/login/generic.py:3]: Unused import.

```suggestion
```

File: fence/config-default.yaml [hunk 4]
```diff
@@ -174,6 +174,19 @@ OPENID_CONNECT:
     # WARNING: DO NOT ENABLE IN PRODUCTION (for testing purposes only)
     mock: false
     mock_default_user: 'test@example.com'
+  # For information on configuring an Okta tenant as an OIDC IdP refer to Okta documentation at:
+  # https://developer.okta.com/docs/reference/api/oidc/#2-okta-as-the-identity-platform-for-your-app-or-api  
+  okta:
```
Comment (williamhaley) [inline fence/config-default.yaml:179]: Can you also add these new sections to `test-fence-config.yaml` for the sake of the automated tests?

File: fence/resources/openid/okta_oauth2.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
+import json
+from .idp_oauth2 import Oauth2ClientBase
+
+OKTA_IDP_NAME = "okta"
```
Comment (williamhaley) [inline fence/resources/openid/okta_oauth2.py:4]: I believe you'll also need to add `okta` to this line in the tests https://github.com/cginmn/fence-service/blob/feat/oidc-client/tests/test_audit_service.py#L366, and to add this...

```
    elif idp == "okta":
        mocked_get_user_id = MagicMock()
        get_user_id_value = {"okta": username}
```

... around here https://github.com/cginmn/fence-service/blob/feat/oidc-client/tests/test_audit_service.py#L336

File: fence/resources/openid/generic_oauth2.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+from .idp_oauth2 import Oauth2ClientBase
+
+GENERIC_IDP_NAME = "generic"
```
Comment (williamhaley) [inline fence/resources/openid/generic_oauth2.py:4]: I believe you'll also need to add `generic` to this line in the tests https://github.com/cginmn/fence-service/blob/feat/oidc-client/tests/test_audit_service.py#L366, and to add this...

```
    elif idp == "generic":
        mocked_get_user_id = MagicMock()
        get_user_id_value = {"generic": username}
```

... around here https://github.com/cginmn/fence-service/blob/feat/oidc-client/tests/test_audit_service.py#L336

**Review by williamhaley — CHANGES_REQUESTED** [review]
Thank you for your contributions! To conform to some of our best practices (apologies that this isn't prominent in our docs) can you please run the following to format your code?

```bash
pip install black
python -m black .
```

Overall I think these changes should work, but I've lever several comments to help our code quality analysis tools and to help get some of the tests to pass. Thank you!

File: fence/resources/openid/generic_oauth2.py (hunk 6 above)
Comment (cginmn) [reply fence/resources/openid/generic_oauth2.py:4]: Sorry i missed all these comments. I forgot about this PR. I opened it some time ago and thought it I had closed it as i had other previous attempts when I was first working with some of the Gen3 folks to understand how to properly get a PR submitted. When i first submitted this one there were a lot of permission issues. I forgot to delete the branch as i had deleted other attempts. This one was replaced eventually by https://github.com/uc-cdis/fence/pull/890. I did incorporate these changes including some code to properly mock the test results on the discovery URL but can also add the additional notes you have. In the 890 PR i did remove the "generic IDP" feature as i was told that was not a design Gen3 folks liked. We have come up with a different more dynamic design but probably need to have a conversation to understand how we can do it better. So the new PR is narrowed down to just Okta.  However i will add those updated to mock user id. 

Let me know if you would like to close this PR or have me close it.

**Comment by cginmn** [comment]
Deleting branch as this was replaced with PR 890.

**Pull request closed without merging**